### PR TITLE
Add the ability to use a custom reader to serve as the random source

### PR DESCRIPTION
**Pull request by jefferai**
This allows for generating repeatable sets of words.

### CODE REVIEW

File: diceware/generate.go [hunk 1]
```diff
@@ -137,21 +149,34 @@ func WordAt(i int) string {
 }
 
 // RollDie rolls a single 6-sided die and returns a value between [1,6].
+//
+// Deprecated: Use Generator.RollDie instead
```
Comment (sethvargo) [inline diceware/generate.go]: What's the goal with deprecating the top-level functions?

File: diceware/generate.go [hunk 2]
```diff
@@ -161,3 +186,11 @@ func RollWord(d int) (int, error) {
 
 	return final, nil
 }
+
+func (g *Generator) RollDie() (int, error) {
```
Comment (sethvargo) [inline diceware/generate.go]: Nit: copy-paste docs please.

File: diceware/generate_test.go [hunk 3]
```diff
@@ -1,7 +1,12 @@
 package diceware
 
 import (
+	"crypto/sha256"
+	"fmt"
+	"reflect"
 	"testing"
+
+	"golang.org/x/crypto/hkdf"
```
Comment (sethvargo) [inline diceware/generate_test.go]: Is there a rand source in the stdlib we can use? This only seems to be for tests, but I really want to rely on no external deps.

File: diceware/generate.go (hunk 1 above)
Comment (jefferai) [reply diceware/generate.go]: I was attempting to follow the example of https://github.com/sethvargo/go-diceware/commit/29ae62910276b023f12faa9be17339ac9475a891 in which you deprecated the top level function in favor of the one on the wordlist. Happy to not mark them deprecated if you prefer.

File: diceware/generate.go (hunk 2 above)
Comment (jefferai) [reply diceware/generate.go]: Oops, yes, sorry.

File: diceware/generate_test.go (hunk 3 above)
Comment (jefferai) [reply diceware/generate_test.go]: Yeah, I'll figure something out. I thought it'd be nicer to test with something representing a real use-case and figured golang/x would be fine, but I can do something else.

File: diceware/generate.go (hunk 1 above)
Comment (sethvargo) [reply diceware/generate.go]: Ah sorry. I was trying to deprecate the top-level wordlist functions, but I want to keep the "default generator" top-level functions.

File: diceware/generate_test.go (hunk 3 above)
Comment (sethvargo) [reply diceware/generate_test.go]: Pulling in from x/crypto is probably "safe", but I like having zero deps

Comment (jefferai) [reply diceware/generate_test.go]: Done.

File: diceware/generate.go (hunk 2 above)
Comment (jefferai) [reply diceware/generate.go]: Done.

File: diceware/generate.go (hunk 1 above)
Comment (jefferai) [reply diceware/generate.go]: Gotcha. Are you okay with what's there so long as it's not marked deprecated?

Comment (sethvargo) [reply diceware/generate.go]: 👍

Comment (jefferai) [reply diceware/generate.go]: Pushed changes

**Comment by sethvargo** [comment]
Thanks!

**Comment by jefferai** [comment]
Thank you!

**Pull request merged**